datafusion/src/scalar.rs

```diff
@@ -526,6 +526,282 @@ macro_rules! eq_array_primitive {
 }
 
 impl ScalarValue {
+    /// Return true if the value is numeric
+    pub fn is_numeric(&self) -> bool {
+        matches!(self,
+            ScalarValue::Float32(_)
+            | ScalarValue::Float64(_)
+            | ScalarValue::Decimal128(_, _, _)
+            | ScalarValue::Int8(_)
+            | ScalarValue::Int16(_)
+            | ScalarValue::Int32(_)
+            | ScalarValue::Int64(_)
+            | ScalarValue::UInt8(_)
+            | ScalarValue::UInt16(_)
+            | ScalarValue::UInt32(_)
+            | ScalarValue::UInt64(_)
+        )
+    }
+
+    /// Add two numeric ScalarValues
+    pub fn add(lhs: &ScalarValue, rhs: &ScalarValue) -> Result<ScalarValue> {
+        if !lhs.is_numeric() || !rhs.is_numeric() {
+            return Err(DataFusionError::Internal(format!(
+                "Addition only supports numeric types, \
+                    here has  {:?} and {:?}",
+                lhs.get_datatype(),
+                rhs.get_datatype()
+            )));
+        }
+
+        // TODO: Finding a good way to support operation between different types without
```

Comment:
Good point, will explore the option.

Reply:
First of all, thanks for suggesting this approach, I was able to get it working after making `evaluate_to_scalar` to `pub(crate)`, it will reduce the amount of code significantly. 

Though after some thinking I am leaning towards having a separate discussion and make the change in a different PR. Here are some of my thoughts:
1. It requires changing the scope of an internal function
2. As you mentioned, there are quite a bit of performance overhead
3. I think ScalarValue arithmetic is important to many future operators, we can take the chance to provide official support. It feels right to formally discuss the topic and review if this is the best way to achieve it. 
4. This method doesn't support uptype when overflow happens. This is useful because for aggregation functions it is quite common the results requires more accuracy than the original type. 
5. Some other functions such as `sum` and `avg` also have similar implementations. Though this is not a good reason to keep as is but it shows the value of official Scalar arithmetic support.

Please let me know your thoughts. If makes sense I am happy to create a new issue for the work and join the discussion. Thanks!

Reply:
> I think ScalarValue arithmetic is important to many future operators,

I think that we should be trying to avoid `ScalarValue` arithmetic if at all possible, and instead use "vectorized" calculations (aka computations using arrow arrays). While this coding style is more cumbersome it is one of the key advantages for datafusion (and columnar execution engines in general)

The reason that `ScalarValue` arithmetic is so slow is that for each row, there is overhead to switch based on the type (aka the large `match` statements required in this PR). Using `Array`s does this type switch once per array so you only pay the cost once for 1000s of rows. 

I think having a separate discussion / PR is a good. 

That being said, I agree that `ScalarValue::sum()` makes more sense that a free `sum` function (as it is easier to find) -- so perhaps as a follow on to this PR we can combine the two implementations